### PR TITLE
implemented profiling API (with do-nothing placeholder version, so it is always safe to call)

### PR DESCRIPTION
**Pull request by boutell**
See #1427 

Placeholder implementation, with tests verifying it does no harm. For now at least the version in the core intentionally does nothing, profiler modules then have an open door to implement it (and the overhead it might involve). The apostrophe-debug module now expects this API.

### CODE REVIEW

File: lib/modules/apostrophe-utils/lib/api.js [hunk 1]
```diff
@@ -622,4 +623,52 @@ module.exports = function(self, options) {
...
+  // `callAll.pageBeforeSend.module-name` as categories. Note that the
+  // most general category should come first.
+  //
+  // To avoid overhead and bloat in the core, the default implementation
+  // does nothing. Also most core modules and methods do not invoke this method.
+  // However, the `apostrophe-profiler` module extends them to invoke it,
+  // for performance reasons: the profiler itself can have
+  // a performance overhead.
+
+  self.profile = function(req, key /* , duration */) {
```
Comment (abea) [inline lib/modules/apostrophe-utils/lib/api.js]: Was this supposed to be uncommented? Duration is referenced on L631, but then the use seems incomplete on L662.

File: lib/modules/apostrophe-utils/lib/api.js [hunk 2]
```diff
@@ -622,4 +623,52 @@ module.exports = function(self, options) {
...
+        return self.profile(null, arguments[0], arguments[1]);
+      } else {
+        return self.profile(null, arguments[0]);
+      }
+    }
+    if (arguments.length === 2) {
+      // This is a do-nothing placeholder implementation
+      return function() {};
+    } else {
+      // All-in-one version with duration, there's nothing to fake
```
Comment (abea) [inline lib/modules/apostrophe-utils/lib/api.js]: I think this should be left out if it's not doing anything yet.

File: lib/modules/apostrophe-utils/lib/api.js [hunk 3]
```diff
@@ -622,4 +623,52 @@ module.exports = function(self, options) {
...
+    }
+  };
+
+  // Returns time since the start of the process in milliseconds,
+  // with high-resolution accuracy. Used by apos.utils.profile.
```
Comment (abea) [inline lib/modules/apostrophe-utils/lib/api.js]: `Used by apos.utils.profile.` is confusing since it doesn't seem to actually be used.

**Comment by boutell** [comment]
@abea please see the description of the PR:

"Placeholder implementation, with tests verifying it does no harm. For now at least the version in the core intentionally does nothing, profiler modules then have an open door to implement it (and the overhead it might involve). The apostrophe-debug module now expects this API."

None of this stuff gets used in the default implementation which is just a placeholder to ensure developers can call it safely without worrying about whether people have chosen to install the forthcoming profiler module or not. For performance reasons a do-nothing placeholder implementation is right for production environments.

**Pull request merged**